core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,62 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.syntax.group._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ */
+case class Perm(mapping: Map[Int, Int]) extends AnyVal {
+  def apply(k: Int): Int = mapping.getOrElse(k, k)
```

Comment:
A permutation is a bijective function (hence the companion `invert` method).
I have half a mind to have this extend `Int => Int` and override `compose` and `andThen`.